presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp

```diff
@@ -139,14 +139,19 @@ folly::SemiFuture<PrestoExchangeSource::Response> PrestoExchangeSource::request(
   // Before calling 'request', the caller should have called
   // 'shouldRequestLocked' and received 'true' response. Hence, we expect
   // requestPending_ == true, atEnd_ == false.
-  // This call cannot be made concurrently from multiple threads.
-  VELOX_CHECK(requestPending_);
-  VELOX_CHECK(!promise_.valid() || promise_.isFulfilled());
+  // This call cannot be made concurrently from multiple threads, but other
+  // calls that mutate members like promise_ can be called concurrently.
+  auto future{folly::SemiFuture<Response>::makeEmpty()};
+  {
+    std::lock_guard<std::mutex> l(queue_->mutex());
+    VELOX_CHECK(requestPending_);
+    VELOX_CHECK(!promise_.valid() || promise_.isFulfilled());
 
-  auto promise = VeloxPromise<Response>("PrestoExchangeSource::request");
-  auto future = promise.getSemiFuture();
+    auto promise = VeloxPromise<Response>("PrestoExchangeSource::request");
```

Comment:
You can create future and promise outside the lock and move to promise_ inside the lock? Is it hard to create a unit test to trigger this race condition? Thanks!

Reply:
True, we can create promise outside the lock. I've move.

I don't know how to create that race. :)
It is probably possible with a lot of thinking and code reading...

Reply:
@xiaoxmeng 
Actually...
If we do this:
```
  VELOX_CHECK(requestPending_);
  auto promise = VeloxPromise<Response>("PrestoExchangeSource::request");
  auto future = promise.getSemiFuture();
  {
    std::lock_guard<std::mutex> l(queue_->mutex());
    VELOX_CHECK(!promise_.valid() || promise_.isFulfilled());
    promise_ = std::move(promise);
  }
```
We might have unfulfilled promise destruction (if VELOX_CHECK() under the lock throws).
Do we care much about creating promise outside of the lock?

Reply:
The unfilled promise won't get destroyed under the lock? We first check promise_ if fails, then we throw. The unfilled promise is destroyed out side the lock? thanks!